From 86788a5ce4da9f0342159e6c78bb1f38508b276f Mon Sep 17 00:00:00 2001
From: Tommy Lin <lin.tommysc@inventec.com>
Date: Wed, 7 Sep 2022 06:37:15 +0000
Subject: [PATCH 28/31] Fix firware version is null issue

Symptom/Reason:
- Firmware Revision is null sometimes after mc reset cold

Root Cause:
- xyz.openbmc_project.Software.Version service was not ready yet
- dev_id_initialized flag will block further accessing once the 1st attempt to get version failed

Solution/Change:
- [phosphor-ipmi-host]
  - add new flag get_version to determine dev_id_initialized flag should be set or not

Entry Test:
- Reset BMC
  - ipmitool mc reset cold
- Get device info
  - ipmitool mc info

---
 apphandler.cpp | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/apphandler.cpp b/apphandler.cpp
index 68e1580..ce72e5f 100644
--- a/apphandler.cpp
+++ b/apphandler.cpp
@@ -929,6 +929,7 @@ ipmi::RspType<uint8_t,  // Device ID
     } devId;
     static bool dev_id_initialized = false;
     static bool defaultActivationSetting = true;
+    static bool get_version = false;
     const char* filename = "/usr/share/ipmi-providers/dev_id.json";
     constexpr auto ipmiDevIdStateShift = 7;
     constexpr auto ipmiDevIdFw1Mask = ~(1 << ipmiDevIdStateShift);
@@ -963,6 +964,7 @@ ipmi::RspType<uint8_t,  // Device ID
             devId.fw[1] = rev.minor % 10 + (rev.minor / 10) * 16;
             std::memcpy(&devId.aux, rev.aux, sizeof(rev.aux));
             haveBMCVersion = true;
+            get_version = true;
         }
     }
 #endif
@@ -1001,7 +1003,7 @@ ipmi::RspType<uint8_t,  // Device ID
                 defaultActivationSetting = data.value("availability", true);
 
                 // Don't read the file every time if successful
-                dev_id_initialized = true;
+                dev_id_initialized = get_version;
             }
             else
             {
-- 
2.25.1